### PR TITLE
Splitting authentication into parallel modules

### PR DESCRIPTION
**Pull request by janosdebugs**
## Please describe the change you are making

This PR splits authentication into parallel modules so that multiple modules can be used on parallel (e.g. webhook for password, oauth for keyboard-interactive).

## Are you the owner of the code you are sending in, or do you have permission of the owner?

Yes

## The code will be published under the MIT-0 license. Have you read and understood this license?

Yes

## TODO

- [ ] Add to audit log
- [x] Make ConnectionMetadata serializable
- [ ] Documentation

### CODE REVIEW

**Comment by janosdebugs** [comment]
TODO: merge metadata for auth and config webhooks in case the server doesn't return any metadata.

**Review by tsipinakis — CHANGES_REQUESTED** [review]

File: internal/auditlog/codec/binary/decode.go
Comment (tsipinakis) [inline internal/auditlog/codec/binary/decode.go:72]: Search and replace error

File: internal/auditlog/storage/file/config.go
Comment (tsipinakis) [inline internal/auditlog/storage/file/config.go:2]: This file can be removed

**Review by tsipinakis — COMMENTED** [review]
Looks like it's coming along really nicely. Though it seems I can't compile it yet. Added some nitpicky comments

File: cmd/containerssh-testauthconfigserver/main.go
Comment (janosdebugs) [reply cmd/containerssh-testauthconfigserver/main.go:62]: We shouldn't use tabs, gofmt will automatically convert to spaces. I'll look into this.

**Review by sanjacodes — APPROVED** [review]
I'd wait for an additional review. You can disregard my comments and just add those changes after this merge in an additional PR.

This PR made me appreciate so many of GitHub's online reviewing functions...

File: auth/webhook/client.go
Comment (bencurio) [inline auth/webhook/client.go:46]: I have discovered a possible validation error in the associated structure.

The problem is in the authentication timeout validation, since `c.Timeout` has been renamed to `c.AuthTimeout`.

Ref link: https://github.com/ContainerSSH/libcontainerssh/blob/a0ceab26028f72e1bc717141c7bef988f881e962/config/auth.go#L409-L411

File: cmd/containerssh-testauthconfigserver/main.go
Comment (janosdebugs) [reply cmd/containerssh-testauthconfigserver/main.go:29]: Let's do that as a separate cleanup PR before release.

File: cmd/swagger/main.go
Comment (janosdebugs) [reply cmd/swagger/main.go:127]: Sadly, this won't work because if the goswagger repo adds a file for checksums that will ruin our code.

File: internal/sshproxy/sshChannelHandler.go
Comment (janosdebugs) [reply internal/sshproxy/sshChannelHandler.go:292]: Removing from this PR as this is not a relevant change.

File: internal/sshproxy/sshConnectionHandler.go
Comment (bencurio) [inline internal/sshproxy/sshConnectionHandler.go:41]: Bad replace?

**Review by sanjacodes — APPROVED** [review]

**Pull request merged**
